pyhap/accessory_driver.py

```diff
@@ -859,118 +861,99 @@ def set_characteristics(self, chars_query, client_addr):
         :type chars_query: dict
         """
         # TODO: Add support for chars that do no support notifications.
-        updates = {}
-        setter_results = {}
-        setter_responses = {}
-        had_error = False
-        had_write_response = False
-        expired = False
 
+        queries = chars_query[HAP_REPR_CHARS]
+
+        self._notify(queries, client_addr)
+
+        updates_by_accessories_services = defaultdict(lambda: defaultdict(dict))
+        results = defaultdict(lambda: defaultdict(dict))
+        char_to_iid = {}
+
+        expired = False
         if HAP_REPR_PID in chars_query:
             pid = chars_query[HAP_REPR_PID]
             expire_time = self.prepared_writes.get(client_addr, {}).pop(pid, None)
-            if expire_time is None or time.time() > expire_time:
-                expired = True
-
-        for cq in chars_query[HAP_REPR_CHARS]:
-            aid, iid = cq[HAP_REPR_AID], cq[HAP_REPR_IID]
-            setter_results.setdefault(aid, {})
+            expired = expire_time is None or time.time() > expire_time
 
-            if HAP_REPR_WRITE_RESPONSE in cq:
-                setter_responses.setdefault(aid, {})
-                had_write_response = True
+        primary_accessory = self.accessory
+        primary_aid = primary_accessory.aid
 
-            if expired:
-                setter_results[aid][iid] = HAP_SERVER_STATUS.INVALID_VALUE_IN_REQUEST
-                had_error = True
+        for query in queries:
+            if HAP_REPR_VALUE not in query and not expired:
                 continue
 
-            if HAP_PERMISSION_NOTIFY in cq:
-                char_topic = get_topic(aid, iid)
-                action = "Subscribed" if cq[HAP_PERMISSION_NOTIFY] else "Unsubscribed"
-                logger.debug(
-                    "%s client %s to topic %s", action, client_addr, char_topic
-                )
-                self.async_subscribe_client_topic(
-                    client_addr, char_topic, cq[HAP_PERMISSION_NOTIFY]
+            aid, iid = query[HAP_REPR_AID], query[HAP_REPR_IID]
+            value = query.get(HAP_REPR_VALUE, None)
```

Comment:
```suggestion
            value = query.get(HAP_REPR_VALUE)
```

Reply:
This one breaks code.

Reply:
Sorry, my mistake. Didn't know get doesn't throw if no default is given.

Reply:
No worries. That's why we have code review 😄